### PR TITLE
Fix OPENJDK-1433, missing FilePermission when reading /etc/pki/java/cacerts with a SM

### PR DESCRIPTION
**Pull request by franferrax**
# Fix [OPENJDK-1433](https://issues.redhat.com/browse/OPENJDK-1433): solr broken due to access denied ("java.io.FilePermission" "/etc/pki/java/cacerts" "read")

This pull request fixes the issue reported in [OPENJDK-1433](https://issues.redhat.com/browse/OPENJDK-1433), and makes some code cleanup (more details about this in the commit messages).

## Reproducer

```java
cat > R.java <<'EOF' && javac R.java && java -Djava.security.manager -cp . R; rm R.*

import javax.net.ssl.SSLContext;

public final class R {
    public static void main(String[] args) throws Throwable {
        SSLContext.getDefault();
        System.out.println("TEST PASS - OK");
    }
}
EOF
```

## Notes

Apart from reviewing the changes against the `cacerts` branch (default PR view), I also recommend giving a look to the [new diff against upstream code](https://github.com/rh-openjdk/jdk8u/compare/jdk8u332-ga...b8d1919c9268e31355dda0a5669e9c2724a0f814#files_bucket), only for `TrustStoreManager.java` and `KeyStoreUtil.java`.

In `git`:

```bash
git diff jdk8u332-ga b8d1919 -- *TrustStoreManager.java *KeyStoreUtil.java
```

```diff
diff --git a/jdk/src/share/classes/sun/security/ssl/TrustStoreManager.java b/jdk/src/share/classes/sun/security/ssl/TrustStoreManager.java
index e7b4763db5..0005e56f52 100644
--- a/jdk/src/share/classes/sun/security/ssl/TrustStoreManager.java
+++ b/jdk/src/share/classes/sun/security/ssl/TrustStoreManager.java
@@ -31,6 +31,7 @@ import java.security.*;
 import java.security.cert.*;
 import java.util.*;
 import sun.security.action.*;
+import sun.security.tools.KeyStoreUtil;
 import sun.security.validator.TrustStoreUtil;
 
 /**
@@ -68,7 +69,7 @@ final class TrustStoreManager {
      * The preference of the default trusted KeyStore is:
      *    javax.net.ssl.trustStore
      *    jssecacerts
-     *    cacerts
+     *    cacerts (system and local)
      */
     private static final class TrustStoreDescriptor {
         private static final String fileSep = File.separator;
@@ -76,7 +77,8 @@ final class TrustStoreManager {
                 GetPropertyAction.privilegedGetProperty("java.home") +
                 fileSep + "lib" + fileSep + "security";
         private static final String defaultStore =
-                defaultStorePath + fileSep + "cacerts";
+                AccessController.doPrivileged((PrivilegedAction<String>) () ->
+                        KeyStoreUtil.getCacertsKeyStorePath());
         private static final String jsseDefaultStore =
                 defaultStorePath + fileSep + "jssecacerts";
 
@@ -139,6 +141,10 @@ final class TrustStoreManager {
                     String storePropPassword = System.getProperty(
                             "javax.net.ssl.trustStorePassword", "");
 
+                    if (SSLLogger.isOn && SSLLogger.isOn("trustmanager")) {
+                        SSLLogger.fine("Default store: " + defaultStore);
+                    }
+
                     String temporaryName = "";
                     File temporaryFile = null;
                     long temporaryTime = 0L;
@@ -160,7 +166,7 @@ final class TrustStoreManager {
                                     SSLLogger.isOn("trustmanager")) {
                                 SSLLogger.fine(
                                         "Inaccessible trust store: " +
-                                        storePropName);
+                                        fileName);
                             }
                         }
                     } else {
diff --git a/jdk/src/share/classes/sun/security/tools/KeyStoreUtil.java b/jdk/src/share/classes/sun/security/tools/KeyStoreUtil.java
index fcc77786da..3a4388964c 100644
--- a/jdk/src/share/classes/sun/security/tools/KeyStoreUtil.java
+++ b/jdk/src/share/classes/sun/security/tools/KeyStoreUtil.java
@@ -41,6 +41,8 @@ import java.text.Collator;
 import java.util.Locale;
 import java.util.ResourceBundle;
 
+import sun.security.util.SecurityProperties;
+
 /**
  * <p> This class provides several utilities to <code>KeyStore</code>.
  *
@@ -54,6 +56,8 @@ public class KeyStoreUtil {
 
     private static final String JKS = "jks";
 
+    private static final String SYSTEM_CA_CERTS_PROP = "security.systemCACerts";
+
     /**
      * Returns true if the certificate is self-signed, false otherwise.
      */
@@ -96,16 +100,30 @@ public class KeyStoreUtil {
         }
     }
 
+    /**
+     * Returns the path to the cacerts DB
+     */
+    public static String getCacertsKeyStorePath()
+    {
+        // Check system DB first, preferring system property over security one
+        String systemDB = SecurityProperties
+                .privilegedGetOverridable(SYSTEM_CA_CERTS_PROP);
+        if (systemDB != null && !"".equals(systemDB) &&
+                (new File(systemDB)).isFile()) {
+            return systemDB;
+        }
+        String sep = File.separator;
+        return System.getProperty("java.home") + sep
+                + "lib" + sep + "security" + sep + "cacerts";
+    }
+
     /**
      * Returns the keystore with the configured CA certificates.
      */
     public static KeyStore getCacertsKeyStore()
         throws Exception
     {
-        String sep = File.separator;
-        File file = new File(System.getProperty("java.home") + sep
-                             + "lib" + sep + "security" + sep
-                             + "cacerts");
+        File file = new File(getCacertsKeyStorePath());
         if (!file.exists()) {
             return null;
         }

```

### CODE REVIEW

**Review by gnu-andrew — APPROVED** [review]
I'm a bit nervous about adding so many additional changes into this one patch, particularly the NPE changes, but after considered review, I think they're safe and the final diff looks a lot better. Thanks for bringing some fresh eyes to the patch.

The checks in `TrustStoreManager.java` date back to when it was handling the user input directly. This was changed in my last revision, though I missed that the `getPath()` could return `null`. [Previously](https://github.com/rh-openjdk/jdk8u/commit/b8da06e1cba2dbbe84a1cabb28d96c342116e230), `TrustStoreManager.java` handled `defaultStore` as upstream and had a `systemStore` direct from user input that could have been `null`.

In other words, this patch has developed some cruft over the years and it's good to clean some of it up now. I guess that's also why we have the duplicate methods; they were moved from `TrustStoreManager.java` without realising that they had been added upstream in 2019.
